### ISSUE TITLE
What's the usage of the for...loop in line 151~152 in 'box_utils'.py?

### ISSUE THREAD

**Issue by feiyuhuahuo** [dbolya/yolact#104]
![q1](https://user-images.githubusercontent.com/32631344/62193897-4f233e00-b3ab-11e9-8a97-2d5d90038bd6.png)
![q2](https://user-images.githubusercontent.com/32631344/62193895-4e8aa780-b3ab-11e9-9b24-051bfba6a53e.png)
The for...loop make 'best_truth_idx' from '[2, 2, 1, 2, 0]' to '[3, 2, 1, 2, 0]', which doesn't make sense. The indexes of the max IoU ground truth box for each prior are still  '[2, 2, 1, 2, 0]' .
Perhaps line 148 should be 'best_prior_overlap.index_fill_(0, best_truth_idx, 2)'? Then all of them make sense. Or I unterstand the code in a wrong way?

**Comment by dbolya** [comment]
Read the above comment:
```python
    # For the best prior for each gt object, set its overlap to 2. This ensures
    # that it won't get thresholded out in the threshold step even if the IoU is
    # under the negative threshold. This is because we want at least one anchor
    # to match with each ground truth or else we'd be wasting training data.
```

Every gt must be matched to a prior regardless of if it's the one with the biggest overlap or not. See in your overlaps array, the y=3 gt has less overlap in every prior than the y=2 gt, so without this step, it wouldn't get used at all (hence `[2, 2, 1, 2, 0]` has no 3). With this step, however, prior x=0 gets assigned to that gt because of the y=3 row, it's the highest overlap (0.4).

Now there's merit in saying this might just confuse the network and shouldn't be done, since the training data that doesn't match any anchor strongly might not be one you want to learn (and I might try training without this rule to see what happens), but this is standard practice in object detectors.

**Comment by feiyuhuahuo** [comment]
![q3](https://user-images.githubusercontent.com/32631344/62202462-ec866e00-b3bb-11e9-928a-c7a92d85d5d8.png)
![q4](https://user-images.githubusercontent.com/32631344/62202463-edb79b00-b3bb-11e9-9592-4eb6924df3a1.png)
I understand this: 'For the best prior for each gt object, set its overlap to 2'.
However, code: 'best_truth_overlap.index_fill_(0, best_prior_idx, 2)' seems not to do this. The 0.4 in the final line should be set to 2 because it's the biggest IoU for line3 (starts from 0). But it just sets the 0.5 in line2 to 2.

**Comment by dbolya** [comment]
So in your example, you have 4 gt objects (y axis) and 5 priors (x axis). The best prior for object 0 (y=0) is x=4 (the 5th one as indicated by 0.8 overlap). The best prior for object 1 (y=1) is similarly x=2 (overlap 0.9). The last two are x=3 and x=0. So, priors x=[0, 2, 3, 4] are given a score of 2 *regardless of what they got matched with*.

You're right in that it's the 0.4 overlap cell that caused x=0 to be chosen, but I'm confused as to why you think the original overlap matrix should be modified? It's not used at all after the max calls. Because of that 0.4, x=0 is chosen to have its score in `best_truth_overlap` set to 2, and indeed that's what happens. Then the for loop you pointed out before sets the best index to that 0.4 cell (x=0, y=3) so it's as if that 0.4 was set to 2.

**Comment by feiyuhuahuo** [comment]
![q5](https://user-images.githubusercontent.com/32631344/62260487-a9bca880-b444-11e9-806f-3745dc5f3148.png)
![q6](https://user-images.githubusercontent.com/32631344/62260488-aaedd580-b444-11e9-9ce4-8115e26490ff.png)
I've understood the mechanism of the for loop. But this method seems still not perfect. The new matrix in the picture has some same element in one row or column. Then the result is wrong. I think the final `best_truth_idx` should be `[3, 2, 1, 0, 1]`, but it was `[3, 2, 1, 2, 1]`, and the gt in line 0 was missed.
Though it's really a small chance for these float numbers to be exactly same, here I'm just discussing the method itself.

**Comment by dbolya** [comment]
The probability for two floating point values to be exactly the same when calculating IoU is so astronomically low that it can be considered as not a problem. However, what worries me with what you bring up is if you changed row 0 column 1 to something less than 0.6, then the same error would still happen.

But then I thought about it, and there are 19k anchors that overtile the entire image. So if two gt bboxes aren't exactly the same, then there will be one anchor with a higher overlap for one, and another with a higher overlap for the other. The only way for this not to happen is for the gap between anchor boxes to perfectly align with the size of the gt box such that an anchor box on either side of this gt box has the same overlap. However, because of the different aspect ratios and scales, I deem this to be almost impossible.

Then finally, the only time this could ever be a problem is if the two gt boxes are exactly the same. Though, I don't know the best course of action if that's the case. I could see someone want to annotate a yin yang symbol like that or something, but an anchor based approach is just not good for that.

**Comment by feiyuhuahuo** [comment]
Sorry, I can not quite understand you. Consider this situation as you mentioned: change the value at row 0 column 1 from 0.6 to 0.4, then the error would still happen. Did you mean that situations like this are almost impossible?

**Comment by dbolya** [comment]
Okay I did some testing with really close gt boxes and it seems that for two 20 x 20 px gt boxes, their centers need to be 6px apart to be assigned to a different anchor box. For bigger boxes (filling the whole screen), they need to be way farther apart, so this is actually a big issue.

It probably won't affect things much because this is only for those gt that didn't match any anchors (which aren't many to begin with hopefully), but I'll push a fix and retrain to see if that improved anything.

**Comment by dbolya** [comment]
I pushed the fix to my test branch and I'll test this on pascal sbd real quick to make sure it didn't break anything. In the mean time, take a look and let me know what you think.

**Comment by feiyuhuahuo** [comment]
I rethink the problem and come with some views. If the overlap matrix is: 
`[[0.1, 0.4, 0.5, 0.6, 0.4],
  [0.3, 0.2, 0.5, 0.8, 0.7],
  [0.5, 0.6, 0.8, 0.9, 0.4],
  [0.4, 0.2, 0.1, 0.2, 0.3]]`, then according your new fix, `best_truth_idx` is `[3, 2, 2, 0, 1]`, and the max IoU priors for gt 1 and gt 2 are missed. The key reason is that there's more than one gt boxes for prior 3 to predict. Then here's my views:
Since what the network actually is doing is that it's predicting 19k anchors, output the offsets for the objects. Thus one anchor could only predict one object. You said that:' since the training data that doesn't match any anchor strongly might not be one you want to learn, but this is standard practice in object detectors.' I don't think let an anchor to learn a low IoU gt box is unsuitable as long as it's the only gt to learn. A low IoU only means that the offsets need to be bigger. But the key problem is that, in testing, if the max IoU anchor for several gt boxes are same, the network must make a judgement for this anchor to decide which gt to predict. This confuses the network, because in training, some anchors are learning their max IoU gt boxes, but some gives up their max IoU gt boxes and are going to learn a lower IoU gt. Like the priors 3 in the above matrix. It learns a 0.6 IoU gt and gives up 0.8 and 0.9 IoU gt.
![1](https://user-images.githubusercontent.com/32631344/62424628-d0f4cd80-b703-11e9-9bd6-770651a4f951.png)

I don't know how many anchors are matching with more than one gt boxes in COCO2017, but I think for small and crowded objects, this situation may come. And if this is a common phenomenon，this may influence the network performance heavily. May be the anchor mechanism is unsuitable for object detectors.

**Comment by dbolya** [comment]
Well that's the point of having 19k anchors. There will always an anchor box relatively close by, and there will always be multiple anchors ready to predict a cluster of multiple got close to each other. Keep in mind that IoU is a very strict metric. An IoU of 0.5 (our threshold) is actually pretty hard to hit, especially for the larger layers. Though I do agree that if we're training the network to jump to a low overlap detection too often, then we'd have problems. I'll see if I can run the statistics 

And about the greedy algorithm causing the earlier got to get priority, that was a conscious decision to keep the speed the same. The alternative would require a second loop and a sort. Though, it's over the number of gt, which is usually small. You know what, I'll just implement the smarter approach then and post an update shortly.

**Comment by dbolya** [comment]
Okay I pushed the updated fix and am testing it now. The new behavior for your most recent example is to return `[3, 2, 0, 2, 1]`, since the `0.9` is attended to first, then the `0.7` of gt 1 (the `0.8` anchor has already been used up by gt 2), etc.

I also ran the stats for what overlaps this new loop is matching (i.e., for each anchor forced to match with a gt because of this loop, I recorded its overlap before I set it to 2). Here's a histogram of the results for the first 10k gt:
![image](https://user-images.githubusercontent.com/19721752/62426978-b9d0c400-b6a0-11e9-8221-98cd73632b27.png)
It looks like most of the matches have IoUs of over around 0.25, with only 8.77% of matches being below an IoU of 0.25 and 2.23% being below 0.1. Now that's not nothing, but I'm thinking that's probably not enough to confuse the network?
